### PR TITLE
Do not record revision if the object is not changed.

### LINKED ISSUE #30030: Revision history erratically loads and removes historic changes
**Describe the bug**
Revision history erratically loads historic changes

**To Reproduce**
Steps to reproduce the behavior (if you can reproduce the bug using the Sample Database, we will find the issue faster):
1. Go to New -> Dashboard -> Edit -> Add a textbox -> Save (notice 2 revisions)

image

2. Edit -> Add a textbox -> Save -> Repeat for more than 8 times --- **Notice how only 7 revisions will show**

image

3. Edit -> Type something in just 1 textbox -> Save (notice that the Revision list went down form 7 to 6)

image

4. Edit -> Type something in just 1 textbox -> Save (repeat this for 7 times)

image

Notice how the revision history vanished. To add with the above You will see that the revision history is always consistent via the  API and it always limits to 14

image

**Expected behavior**
Consistency with the API

**Information about your Metabase Installation:**

Happens in 1.46.1 and `master` 

**Additional context**

Please note that in master for some reason only 4 revisions show up

### PR DESCRIPTION
**Pull request by qnkhuat**
Fixes #30030

The change is to make sure we only record a revision only when the object changes. (Check this comment for the reasoning https://github.com/metabase/metabase/issues/30030#issuecomment-1517629594. )

I think this is safe and it should be implemented a long time ago.

In fact, the way we display revision on FE is kinda a hack. Currently, if you edit the dashboard, we record at least 3 revisions with the same object, but if you open the details on FE, it only says there is one revision.
The reason for this is FE intelligently does a diff between revisions to filter out only the one that changes... not sure why we did it that way but I think it should be implemented in the BE from start.

---
Product also can't think of any reasons why we need revision without changes ([thread](https://metaboat.slack.com/archives/C01LQQ2UW03/p1682079144330459?thread_ts=1682078063.301779&cid=C01LQQ2UW03))

This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/30285)

### CODE REVIEW

File: src/metabase/models/revision.clj [hunk 1]
```diff
@@ -134,19 +135,21 @@
          (integer? id)
          (t2/exists? entity :id id)
          (map? object)]}
-  (let [object (serialize-instance entity id (dissoc object :message))]
+  (let [serialized-object (serialize-instance entity id (dissoc object :message))
+        last-object       (t2/select-one-fn :object Revision :model (name entity) :model_id id {:order-by [[:id :desc]]})]
```
Comment (calherries) [inline src/metabase/models/revision.clj:139]: This works, but it's a waste of a DB call if we already know that the object has changed or not further up the call tree. For example we should be able to skip this check if `is-creation?` is true, or if it's a delete event. I also checked the call sites for `:dashboard-update`, `:segment-update`, and `:metric-update`, and it seems like we should be able to calculate whether or not anything has actually changed there too.

**Review by calherries — CHANGES_REQUESTED** [review]
The solution works to solve the issue, but it introduces an extra DB call and in many cases it is not needed. I think checking whether a revision has occurred should be made further up the call stack, because this seems to only be needed for update revisions, not create or delete revisions, and in many cases where these events are published we should know already whether something has actually been updated.

To me the real problem that should be solved is that a `:dashboard-update` event is published even though the dashboard has not actually changed.

**Comment by calherries** [comment]
To show you what I mean, here's a patch that makes sure the `:dashboard-update` event is only published if the dashboard has actually changed:

[0001-Only-update-dashboard-if-changes-have-happened.patch]
(https://github.com/metabase/metabase/files/11309958/0001-Only-update-dashboard-if-changes-have-happened.patch)

It should solve the original issue too.

**Comment by qnkhuat** [comment]
I'm not sure saving 1 DB call is really worth the complication. This is not only around `:dashboard-update` but also, `:dashboard-reposition-cards` too.

Also this happens async, so I think handling it at low level is both cleaner + more powerful.

**Review by calherries — APPROVED** [review]

**Pull request merged**